klippa/src/hvar.rs

```diff
@@ -0,0 +1,640 @@
+                if v.inner > this.max_inners[outer] {
+                    this.max_inners[outer] = v.inner;
+                }
+
+                inner_sets[outer].insert(v.inner);
+            }
+        }
+
+        Ok(this)
+    }
+
+    fn remap(
+        &mut self,
+        plan: &Plan,
+        index_map: &Option<DeltaSetIndexMap>,
+        outer_map: &IncBiMap,
+        inner_maps: &[IncBiMap],
+    ) {
+        self.inner_bit_count = 1;
+
+        for (max_inner, inner_map) in self.max_inners.iter().zip(inner_maps) {
+            if inner_map.len() == 0 || *max_inner == 0 {
+                continue;
+            }
+
+            let bit_count = 32
+                - inner_map
+                    .get(*max_inner as u32)
+                    .unwrap_or(&0)
+                    .leading_zeros() as u8;
+            self.inner_bit_count = bit_count.max(self.inner_bit_count);
+        }
+
+        for (new_gid, old_gid) in plan.new_to_old_gid_list.iter() {
+            if new_gid.to_u32() as u16 >= self.map_count {
+                break;
+            }
+
+            let old_gid = old_gid.to_u32();
+            let v = match index_map {
+                Some(m) => m.get(old_gid).unwrap(),
+                None => DeltaSetIndex {
+                    outer: (old_gid >> 16) as u16,
+                    inner: (old_gid & 0xFFFF) as u16,
+                },
+            };
+
+            let outer = v.outer;
+            if outer as usize >= inner_maps.len() {
+                continue;
+            }
+
+            let new_outer = outer_map.get(outer as u32).unwrap();
+            let new_inner = inner_maps[outer as usize].get(v.inner as u32).unwrap();
+            self.output_map
+                .insert(*new_gid, (*new_outer << 16) | *new_inner);
+        }
+    }
+
+    fn is_identity(&self) -> bool {
+        self.output_map.is_empty()
+    }
+
+    fn to_serialize_plan(&self) -> DeltaSetIndexMapSerializePlan {
+        DeltaSetIndexMapSerializePlan::new(
+            self.outer_bit_count,
+            self.inner_bit_count,
+            &self.output_map,
+            self.map_count as u32,
+        )
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct HvarVvarSubsetPlan {
+    inner_maps: Vec<IncBiMap>,
+    index_map_subset_plans: Vec<IndexMapSubsetPlan>,
+}
+
+impl HvarVvarSubsetPlan {
+    pub(crate) fn new(
+        plan: &Plan,
+        var_store: &ItemVariationStore,
+        index_maps: &[Option<DeltaSetIndexMap>],
+    ) -> Result<Self, ReadError> {
+        let mut this = HvarVvarSubsetPlan::default();
+        let vardata_count = var_store.item_variation_data_count() as usize;
+
+        let mut inner_sets = Vec::new();
+        inner_sets.resize(vardata_count, Default::default());
+
+        let mut outer_map = IncBiMap::default();
+        // don't bypass empty adv map
+        let index_map_subset_plan =
+            IndexMapSubsetPlan::init(&index_maps[0], plan, false, &mut outer_map, &mut inner_sets)?;
```

Comment:
is there a particular reason for passing these in? In general I would prefer to construct them in the called method, and return a tuple if needed...

Reply:
oh it's because these are more global and would need to be updated in each constructor call, we cannot just construct them and return